web/src/message_store.ts

```diff
@@ -403,19 +409,361 @@ export function reify_message_id({old_id, new_id}: {old_id: number; new_id: numb
+            stream_id: to_stream_id,
+            topic: to_topic,
+            message_id: to_message_id,
+        });
+        topic_links_from_message_narrow.set(message.id, topic_links_from_message);
+
+        // Update topic_links_by_to if relevant
+        if (to_topic !== undefined) {
+            const topic_links_to_message_narrow = get_or_create_link_map_for_narrow(
+                topic_links_by_to,
+                to_stream_id,
+                to_topic,
+            );
+            const topic_links_to_message =
+                topic_links_to_message_narrow.get(to_message_id ?? NO_MESSAGE_ID) ?? [];
+            // Note: This is O(# of links for message_id, which could be a lot for NO_MESSAGE_ID)
+            unique_array_insert(topic_links_to_message, message.id);
+            topic_links_to_message_narrow.set(
+                to_message_id ?? NO_MESSAGE_ID,
+                topic_links_to_message,
+            );
+        }
+    }
+}
+
+// Important: Messages should still have the old stream and topic,
+// so that we can fetch them in the link maps.
+export function process_topic_edit(opts: {
+    message_ids: number[];
+    new_stream_id: number;
+    new_topic: string;
+}): void {
+    const {message_ids, new_stream_id, new_topic} = opts;
+    for (const message_id of message_ids) {
+        const message = get(message_id);
+        assert(message?.type === "stream");
+
+        const links_from_old_narrow = topic_links_by_from
+            .get(message.stream_id)
+            ?.get(message.topic);
+        const links_from_edited_message = links_from_old_narrow?.get(message_id);
+        // Move any links from this message stored with the old topic
+        if (links_from_edited_message !== undefined) {
+            const new_narrow_link_map = get_or_create_link_map_for_narrow(
+                topic_links_by_from,
+                new_stream_id,
+                new_topic,
+            );
+            new_narrow_link_map.set(message_id, links_from_edited_message);
+            links_from_old_narrow!.delete(message_id);
+        }
+
+        // Move any links to this message stored with the old topic
+        const links_to_old_narrow = topic_links_by_to.get(message.stream_id)?.get(message.topic);
+        const links_to_edited_message = links_to_old_narrow?.get(message_id);
+        if (links_to_edited_message !== undefined) {
+            const new_narrow_link_map = get_or_create_link_map_for_narrow(
+                topic_links_by_to,
+                new_stream_id,
+                new_topic,
+            );
+            new_narrow_link_map.set(message_id, links_to_edited_message);
+            links_to_old_narrow!.delete(message_id);
+        }
+    }
+}
+
+export function update_message_content(
+    message: Message,
+    new_content: string,
+    may_have_new_links = true,
+): void {
     message.content = new_content;
+    if (message.type !== "stream" || !may_have_new_links) {
+        return;
+    }
```

Comment:
This block probably serves a comment on why DMs aren't included, possibly with a (not a priority) "TODO" note about what might make sense to do in DMs. (I think outgoing links from DM conversations could be useful to track, but would require new data structures?).

Also, is `may_have_new_links` the right check? I'd expect removing links to be relevant too...